### ISSUE TITLE
curl can't authenticate using --negotiate to Kerberos or NTLM sites

### ISSUE THREAD

**Issue by NadeemAfana** [curl/curl#980]
#### Cannot authenticate to Kerberos or NTLM using --negotiate. I get the following error:
**`gss_init_sec_context() failed: SPNEGO cannot find mechanisms to negotiate.`**

**Note**: This was working for version 7.30

### Output

`curl -v  --negotiate -u : http://mysite:8000/test.aspx`
```
* STATE: INIT => CONNECT handle 0x6000608a0; line 1396 (connection #-5000)
* Added connection 0. The cache now contains 1 members
*   Trying 10.45.159.215...
* STATE: CONNECT => WAITCONNECT handle 0x6000608a0; line 1449 (connection #0)
* Connected to mysite (10.45.159.215) port 8000 (#0)
* STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x6000608a0; line 1556 (connection #0)
* STATE: SENDPROTOCONNECT => DO handle 0x6000608a0; line 1574 (connection #0)
> GET /test.aspx HTTP/1.1
> Host: mysite:8000
> User-Agent: curl/7.50.1
> Accept: */*
>
* STATE: DO => DO_DONE handle 0x6000608a0; line 1653 (connection #0)
* STATE: DO_DONE => WAITPERFORM handle 0x6000608a0; line 1780 (connection #0)
* STATE: WAITPERFORM => PERFORM handle 0x6000608a0; line 1790 (connection #0)
* HTTP 1.1 or later with persistent connection, pipelining supported
< HTTP/1.1 401 Unauthorized
< Content-Type: text/html
* Server  is not blacklisted
< Server:
* gss_init_sec_context() failed: SPNEGO cannot find mechanisms to negotiate.
< WWW-Authenticate: Negotiate
< WWW-Authenticate: NTLM
< X-Frame-Options: SAMEORIGIN
< Date: Wed, 24 Aug 2016 00:37:28 GMT
< Content-Length: 1293
<
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;}
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;}
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;}
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>
* STATE: PERFORM => DONE handle 0x6000608a0; line 1949 (connection #0)
* multi_done
* Connection #0 to host mysite left intact
* Expire cleared

```

### curl/libcurl version
```
curl --version

curl 7.50.1 (x86_64-unknown-cygwin) libcurl/7.50.1 OpenSSL/1.0.2h zlib/1.2.8 libidn/1.29 libpsl/0.13.0 (+libidn/1.29) libssh2/1.7.0 nghttp2/1.7.1
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: Debug IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets Metalink PSL
```

### operating system
Windows 10

**Comment by frenche** [comment]
Looks like curl was built with gssapi support instead of windows sspi.
Are you sure that's what you want? In that case you may need to get initial ticket first (kinit?).

**Comment by NadeemAfana** [comment]
There is no builtin kinit in Windows. Klist is used to get tickets instead.  I used klist to get the corresponding service ticket but still getting same error from curl.





From: Isaac Boukris

Sent: Wednesday, September 14, 4:08 PM

Subject: Re: [curl/curl] curl can't authenticate using --negotiate to Kerberos or NTLM sites (#980)

To: curl/curl

Cc: Nadeem Afana, Author



Looks like curl was built with gssapi support instead of windows sspi.

Are you sure that's what you want? In that case you may need to get initial ticket first (kinit?).

**Comment by frenche** [comment]
If you have GSSAPI library (like MIT Kerberos for windows) then you should have 'kinit'.
The 'klist' you have on windows if from the native MS Kerberos (of the windows system).
If you want to use MS SSPI back-end, then you should rebuild curl against it (not with GSSAPI).

**Comment by NadeemAfana** [comment]
Thanks for clarifying. Is there a compiled binary available for download? I can only find verson 7.33 at https://curl.haxx.se/latest.cgi

**Comment by frenche** [comment]
Perhaps try here: https://curl.haxx.se/download.html
I'd try the generic windows ones ('curl -V' should indicate SSPI instead of GSSAPI).
